File: UsoguiLabyrinth/Labyrinth.cpp
#include <iostream>
#include <cmath>
#include <string>
#include <stdlib.h>
#include <iterator>
#include <chrono>

#include "Labyrinth.h"
using std::cout;
using std::cin;
using std::endl;
using std::string;
using std::make_pair;
using std::chrono::system_clock;
using std::chrono::duration_cast;
using std::chrono::milliseconds;
using std::chrono::duration;
using std::chrono::time_point;

#define _UP_ 'w'
#define _DOWN_ 's'
#define _LEFT_ 'a'
#define _RIGHT_ 'd'
#define _VWALL_ '|'
#define _HWALL_ '='
#define _JOINT_ '+'
#define _ROOM_ ' '
#define UP 1
#define DOWN 3
#define RIGHT 2
#define LEFT 4
#define UPPER_LIMIT 2
#define LEFT_LIMIT 2
#define RIGHT_LIMIT 12
#define DOWN_LIMIT 12
#define SIZE 6

#ifndef DEBUG(x, y)
#define DEBUG(x, y) cout << "[DEBUG] " << x << y << endl; (_getch() != char(13) ? _getch() : 0);
#endif // DEBUG(x, y) cout << x << y << endl; (_getch() != char(13) ? _getch() : 0);

#ifndef PAUSE
#define PAUSE if (int c = _getch() != char(13)) {c = _getch();}
#endif // !1

#ifndef RANDOM_SEED(x)
#define RANDOM_SEED(x) duration_cast<milliseconds>(x.time_since_epoch()).count()
#endif // !1

#ifndef DISTANCE(x, y)
#define DISTANCE(x, y) sqrt(pow(start.first - finish.first, 2)+pow(start.second - finish.second, 2))
#endif // ! DISTANCE(x, y) sqrt(pow(start.first - finish.first, 2)+pow(start.second - finish.second, 2))

Labyrinth::Labyrinth() : n_walls(0){
	char side = 'A', column = '1';
	vector<char> row;

	row.push_back(_ROOM_);
	row.push_back(_VWALL_);
	for (int i = 0; i < (SIZE * 2); ++i) {
		row.push_back(column);
		row.push_back(_ROOM_);
		++column;
	}
	map.push_back(row);
	row.clear();

	row.push_back(_HWALL_);
	row.push_back(_JOINT_);
	for (int i = 0; i < (SIZE * 2); ++i) {
		row.push_back(_HWALL_);
	}
	map.push_back(row);
	row.clear();

	for (int i = 0; i < (SIZE * 2); ++i) {
		if (i % 2 == 0) {
			row.push_back(side);
			++side;
			row.push_back(_VWALL_);
			for (int k = 0; k < SIZE; ++k) {
				row.push_back(_ROOM_);
				row.push_back(_ROOM_);
			}
		}
		else {
			row.push_back(_ROOM_);
			row.push_back(_VWALL_);
			for (int k = 0; k < SIZE; ++k) {
				row.push_back(_ROOM_);
				row.push_back(_JOINT_);
			}
		}

		map.push_back(row);
		row.clear();
	}
}

void Labyrinth::printLabyrinth(const string& name, bool show_walls) {
	cout << "Player: " << name << endl;
	char c;
	for (int i = 0; i < 14; ++i) {
		for (int j = 0; j < 14; ++j) {
			c = map.at(i).at(j);
			cout << ((i >= 2 && j >= 2) && (c == _VWALL_ || c == _HWALL_) && !show_walls ? ' ' : c) << std::flush;
		}
		cout << endl;
	}
}

coordinate Labyrinth::translate(coordinate c) const {
	c.first = (c.first * 2) + 2;
	c.second = (c.second * 2) + 2;
	return c;
}

char Labyrinth::at(coordinate pos) const {
	pos = translate(pos);
	return map.at(pos.first).at(pos.second);
}

void Labyrinth::alter(coordinate pos, char c) {
	pos = translate(pos);
	map.at(pos.first).at(pos.second) = c;
	return;
}

bool Labyrinth::BuildWall(coordinate pos, int dir) {
	pos = translate(pos);
	
	switch (dir) {
	case UP:
		--pos.first;
		break;
	case RIGHT:
		++pos.second;
		break;
	case DOWN:
		++pos.first;
		break;
	case LEFT:
		--pos.second;
		break;
	}

	if (map.at(pos.first).at(pos.second) != _ROOM_)
		return false;
	if (dir == 1 || dir == 3) {
		map.at(pos.first).at(pos.second) = _HWALL_;
	}
	else if (dir == 2 || dir == 4) {
		map.at(pos.first).at(pos.second) = _VWALL_;
	}
	this->AddDelWall(1);
	return true;
}

bool Labyrinth::EraseWall(coordinate pos, int dir) {
	pos = translate(pos);

	switch (dir) {
	case UP:
		--pos.first;
		break;
	case RIGHT:
		++pos.second;
		break;
	case DOWN:
		++pos.first;
		break;
	case LEFT:
		--pos.second;
		break;
	}

	if ((pos.first <= 2 && dir == 1) || (pos.second <= 2 && dir == 4))
		return false;
	
	map.at(pos.first).at(pos.second) = ' ';
	this->AddDelWall(-1);
	return true;
}

bool Labyrinth::isClosed(coordinate pos, char dir) {
	pos = translate(pos);
	switch (dir) {
	case _UP_:
		return (map.at(pos.first - 1).at(pos.second) != _ROOM_ && pos.first > UPPER_LIMIT ? true : false);
	case _DOWN_:
		return (map.at(pos.first + 1).at(pos.second) != _ROOM_ && pos.first < DOWN_LIMIT ? true : false);
	case _LEFT_:
		return (map.at(pos.first).at(pos.second - 1) != _ROOM_ && pos.second > LEFT_LIMIT ? true : false);
	case _RIGHT_:
		return (map.at(pos.first).at(pos.second + 1) != _ROOM_ && pos.second < RIGHT_LIMIT ? true : false);
	default:
		return false;
	}
}

coordinate Labyrinth::getStart() const {
	coordinate c = this->start;
	c.first = (c.first / 2) - 1;
	c.second = (c.second / 2) - 1;
	return c;
}

coordinate Labyrinth::getFinish() const {
	coordinate c = this->finish;
	c.first = (c.first / 2) - 1;
	c.second = (c.second / 2) - 1;
	return c;
}

coordinate Labyrinth::SetupLabyrinth() {
	time_point<system_clock> t = system_clock::now();
	srand(RANDOM_SEED(t));
	coordinate start, finish;

	for (int i = 0; i < 6; ++i) {
		for (int j = 0; j < 6; ++j) {
			this->BuildWall(make_pair(i, j), 1);
			this->BuildWall(make_pair(i, j), 4);
			if (i < 5)
				this->BuildWall(make_pair(i, j), 3);
			if (j < 5)
				this->BuildWall(make_pair(i, j), 2);
		}
	}

	do {
		start.first = rand() % 6;
		start.second = rand() % 6;
		finish.first = rand() % 6;
		finish.second = rand() % 6;
	} while (DISTANCE(start, finish) < 4.75);

	this->setStart(start);
	this->setFinish(finish);
	BuildLabyrinth();
	return start;
}

void Labyrinth::BuildLabyrinth() {
	int dir;
	coordinate pos, temp;
	Grid grid;

	pos.first = rand() % 6;
	pos.second = rand() % 6;

	grid.Open(pos);
	while (grid.getClosed().size() > 0) {
		dir = 1 + rand() % 4;
		temp = pos;
		switch (dir % 2)
		{
		case 1:
			temp.first += (dir == 1 ? -1 : 1);
			break;
		case 0:
			temp.second += (dir == 2 ? 1 : -1);
			break;
		default:
			break;
		}
		if (std::find(grid.getFrontier().begin(), grid.getFrontier().end(), temp) != grid.getFrontier().end()) {
			grid.Open(temp);
			this->EraseWall(pos, dir);
		}

		pos = grid.getRandomOpen();
	}
}

void Labyrinth::setStart(coordinate pos) {
	this->start = translate(pos);
	this->alter(pos, 'S');
}

void Labyrinth::setFinish(coordinate pos) {
	this->finish = translate(pos);
	this->alter(pos, 'F');
}

void Labyrinth::AddDelWall(int condition) {
	this->n_walls += condition;
}

int Labyrinth::getNofWalls() const {
	return this->n_walls;
}

Room::Room() : open(false), frontier(false), closed(true) {

}

Grid::Grid() {
	for (int i = 0; i < 6; ++i) {
		for (int j = 0; j < 6; ++j) {
			closed.push_back(make_pair(i, j));
		}
	}
}

int Grid::Open(coordinate pos) {
	rooms[pos.first][pos.second].open = true;
	rooms[pos.first][pos.second].closed = false;
	rooms[pos.first][pos.second].frontier = false;

	this->open.push_back(make_pair(pos.first, pos.second));
	this->frontier.remove(pos);
	this->closed.remove(pos);

	// else-if not used bcs they function separately
	if (pos.first > 0 && rooms[pos.first - 1][pos.second].closed == true) {
		rooms[pos.first - 1][pos.second].frontier = true;
		this->frontier.push_back(make_pair(pos.first - 1, pos.second));
	}
	if (pos.first < 5 && rooms[pos.first + 1][pos.second].closed == true) {
		rooms[pos.first + 1][pos.second].frontier = true;
		this->frontier.push_back(make_pair(pos.first + 1, pos.second));
	}
	if (pos.second > 0 && rooms[pos.first][pos.second - 1].closed == true) {
		rooms[pos.first][pos.second - 1].frontier = true;
		this->frontier.push_back(make_pair(pos.first, pos.second - 1));
	}
	if (pos.second < 5 && rooms[pos.first][pos.second + 1].closed == true) {
		rooms[pos.first][pos.second + 1].frontier = true;
		this->frontier.push_back(make_pair(pos.first, pos.second + 1));
	}

	return this->open.size();
}

const list<coordinate>& Grid::getOpen() const {
	return this->open;
}

const coordinate& Grid::getRandomOpen() const {
	int index = rand() % this->open.size();
	list<coordinate>::const_iterator it = this->open.begin();
	std::advance(it, index);
	return *it;
}

const list<coordinate>& Grid::getClosed() const {
	return this->closed;
}

const list<coordinate>& Grid::getFrontier() const {
	return this->frontier;
}

File: java/Labyrinth/src/Labyrinth.java
/*
 * It represents the labyrinth. Each Player object contains one. 
 * It also allows for other classes to alter a labyrinth by providing appropriate methods
 */

import java.util.*;

class Labyrinth
{
	public static boolean inner = false; //flag for when one of its own methods calls the Alter
	private int n_walls;				 //number of walls
	private coordinate start, finish;	 //starting and finishing point
	private Player player;				 //the player that is either making the labyrinth or going to solve it
	private GameScreen game;			 //the window that will show the labyrinth to the player
	private ArrayList<ArrayList<Character>> map; //a matrix that internally represents the labyrinth

	private coordinate translate(coordinate c) { 
		//the other classes and this one refer to the same points on the labyrinth with different values
		//so a function was needed to translate the values from the external set to these of the internal
		coordinate temp = new coordinate(c);
		temp.y_axis = (temp.y_axis * 2) + 2;
		temp.x_axis = (temp.x_axis * 2) + 2;
		return temp;
	}
	
	public Labyrinth(GameScreen p_game) { //it sets the labyrinth on its initial state with only the very basic walls
		this.game = p_game;
		n_walls = 0;
		map = new ArrayList<ArrayList<Character>>();
		start = new coordinate();
		finish = new coordinate();
		char side = 'A', column = '1';
		ArrayList<Character> row = new ArrayList<Character>(); //it first readies a row and then add it to actual labyrinth

		//the first row that contains the numbers above is added
		row.add(' ');
		row.add('|');
		for (int i = 0; i < 6; ++i) {
			row.add(column);
			row.add(' ');
			++column;
		}
		map.add(new ArrayList<Character>(row));
		row.clear();

		//the second row that contains the upper walls to separate the top row from the labyrinth is added
		row.add('=');
		row.add('+');
		for (int i = 0; i < (6 * 2); ++i) {
			row.add('=');
		}
		map.add(new ArrayList<Character>(row));
		row.clear();

		//the rest of the rows are made and added
		for (int i = 0; i < (6 * 2); ++i) {
			if (i % 2 == 0) {
				row.add(side); //it adds the letter on the far left of some rows
				++side;
				row.add('|');
				for (int k = 0; k < 6; ++k) {
					row.add(' ');
					row.add(' ');
				}
			}
			else {
				row.add(' ');
				row.add('|');
				for (int k = 0; k < 6; ++k) {
					row.add(' ');
					row.add('+');
				}
			}
			map.add(new ArrayList<Character>(row));
			row.clear();
		}
	}

	public char at(coordinate pos) { //return the character representing the sate of a point in the labyrinth
		coordinate temp = (translate(pos));
		return map.get(temp.y_axis).get(temp.x_axis);
	}
	
	public char ingameAt(coordinate pos) { 
		//return the character representing the sate of a point in the labyrinth for when an internal method needs it
		//so no value translation is needed
		return map.get(pos.y_axis).get(pos.x_axis);
	}
	
	public void alter(coordinate pos, char c) { //changes the state of a point in labyrinth
		//translates the position values given if the call isn't made from an internal method
		coordinate temp = (inner) ? (pos) : translate(pos);
		map.get(temp.y_axis).set(temp.x_axis, c);
		temp = (!inner) ? translate(pos) : temp;
		game.Alter(temp, c);
		inner = false;
		return;
	}
	
	public boolean BuildWall(coordinate pos, int dir) {//it adds a wall given a position and a direction
		coordinate temp = (translate(pos));
		
		switch (dir) {
		case 1://Up
			--temp.y_axis;
			break;
		case 2://Right
			++temp.x_axis;
			break;
		case 3://Down
			++temp.y_axis;
			break;
		case 4://Left
			--temp.x_axis;
			break;
		}

		if (!map.get(temp.y_axis).get(temp.x_axis).equals(' '))
			return false;
		if (map.get(temp.y_axis).get(temp.x_axis).equals(' '))
			this.AddDelWall(1);
		if (dir == 1 || dir == 3) {
			inner = true;
			alter(temp, '=');
		}
		else if (dir == 2 || dir == 4) {
			inner = true;
			alter(temp, '|');
		}
		return true;
	}
	
	public boolean EraseWall(coordinate pos, int dir) {//it removes a wall given a position and a direction
		coordinate temp = (translate(pos));

		switch (dir) {
		case 1://Up
			--temp.y_axis;
			break;
		case 2://Right
			++temp.x_axis;
			break;
		case 3://Down
			++temp.y_axis;
			break;
		case 4://Left
			--temp.x_axis;
			break;
		}

		if ((temp.y_axis <= 2 && dir == 1) || (temp.x_axis <= 2 && dir == 4))
			return false;
		if (!map.get(temp.y_axis).get(temp.x_axis).equals(' '))
			this.AddDelWall(-1);
		inner = true;
		alter(temp, ' ');
		return true;
	}
	public boolean isClosed(coordinate pos, char dir) {
		//tells the caller class if there is wall somewhere given a position and a direction
		coordinate temp = translate(pos);
		switch (dir) {
		case 'w'://Up
			return (!map.get(temp.y_axis - 1).get(temp.x_axis).equals(' ') && temp.y_axis > 2 ? true : false);
		case 's'://Down
			return (!map.get(temp.y_axis + 1).get(temp.x_axis).equals(' ') && temp.y_axis < 12 ? true : false);
		case 'a'://Left
			return (!map.get(temp.y_axis).get(temp.x_axis - 1).equals(' ') && temp.x_axis > 2 ? true : false);
		case 'd'://Right
			return (!map.get(temp.y_axis).get(temp.x_axis + 1).equals(' ') && temp.x_axis < 12 ? true : false);
		default:
			return false;
		}
	}
	public coordinate getStart() {//returns staring point
		coordinate c = new coordinate(this.start);
		c.y_axis = (c.y_axis / 2) - 1;
		c.x_axis = (c.x_axis / 2) - 1;
		return c;
	}
	public coordinate getFinish() {//returns finishing point
		coordinate c = new coordinate(this.finish);
		c.y_axis = (c.y_axis / 2) - 1;
		c.x_axis = (c.x_axis / 2) - 1;
		return c;
	}
	void SetupLabyrinth() {//makes basic needed preparations for the labyrinth to be randomly generated
		Random t = new Random();
		coordinate start = new coordinate(), finish = new coordinate();

		for (int i = 0; i < 6; ++i) {
			for (int j = 0; j < 6; ++j) {
				this.BuildWall(coordinate.make_pair(i, j), 1);
				this.BuildWall(coordinate.make_pair(i, j), 4);
				if (i < 5)
					this.BuildWall(coordinate.make_pair(i, j), 3);
				if (j < 5)
					this.BuildWall(coordinate.make_pair(i, j), 2);
			}
		}
		
		//System.out.println("Checking distance");
		do {
			start.y_axis = t.nextInt(6);
			start.x_axis = t.nextInt(6);
			finish.y_axis = t.nextInt(6);
			finish.x_axis = t.nextInt(6);
		} while (Math.sqrt(Math.pow(start.y_axis - finish.y_axis, 2)+Math.pow(start.x_axis - finish.x_axis, 2)) < 4.75);

		this.setStart(start);
		this.setFinish(finish);
	}
	public coordinate GenerateLabyrinth() {//randomly generates a labyrinth
		int dir;
		coordinate pos = new coordinate(), temp = new coordinate();
		Grid grid = new Grid();
		Random t = new Random();

		SetupLabyrinth();
		
		pos.y_axis = t.nextInt(6);
		pos.x_axis = t.nextInt(6);

		grid.Open(pos);
		while (grid.getClosed().size() > 0) {
			dir = 1 + t.nextInt(4);
			temp.copy(pos);
			switch (dir % 2)
			{
			case 1:
				temp.y_axis += (dir == 1 ? -1 : 1);
				break;
			case 0:
				temp.x_axis += (dir == 2 ? 1 : -1);
				break;
			default:
				break;
			}
			if (grid.getFrontier().contains(temp)) {
				grid.Open(temp);
				this.EraseWall(pos, dir);
			}

			pos.copy(grid.getRandomOpen());
		}

		while (n_walls > 20) {
			dir = 1 + t.nextInt(4);
			pos = grid.getRandomOpen();
			this.EraseWall(pos, dir);
		}

		return this.getStart();
	}
	public void setStart(coordinate pos) {//sets starting point
		this.start.copy(translate(pos));
		inner = false;
		this.alter(pos, 'S');
	}
	public void setFinish(coordinate pos) {//sets finishing point
		this.finish.copy(translate(pos));
		inner = false;
		this.alter(pos, 'f');
	}
	private void AddDelWall(int condition) {//increases or decreases the recorded number of walls
		this.n_walls += condition;
	}
	public int getNofWalls() {//returns the number of walls
		return this.n_walls;
	}
	
	public void setGameScreen(GameScreen p_game) {//sets the window that the labyrinth will be shown
		this.game = p_game;
	}
	
	public void SetPlayer(Player p_player) {//sets the player that is either making the labyrinth or going to solve it
		player = p_player;
	}
	
	public Player getPlayer() {//returns the player that is either making the labyrinth or going to solve it
		return player;
	}
};
